homeassistant/components/sonos/alarms.py

```diff
@@ -0,0 +1,80 @@
+"""Class representing Sonos alarms."""
+from __future__ import annotations
+
+from collections import deque
+from collections.abc import Iterator
+import logging
+from typing import Callable
+
+from pysonos import SoCo
+from pysonos.alarms import Alarm, get_alarms
+from pysonos.events_base import Event as SonosEvent
+from pysonos.exceptions import SoCoException
+
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.dispatcher import async_dispatcher_send
+
+from .const import DATA_SONOS, SONOS_ALARM_UPDATE, SONOS_CREATE_ALARM
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class SonosAlarms:
+    """Storage class for Sonos alarms."""
+
+    def __init__(self, hass: HomeAssistant, household_id: str) -> None:
+        """Initialize the data."""
+        self.hass = hass
+        self.household_id = household_id
+        self._alarms: dict[str, Alarm] = {}
+        self._next_update: Callable | None = None
+        self._processed_events = deque(maxlen=5)
+
+    def __iter__(self) -> Iterator:
+        """Return an iterator for the known alarms."""
+        alarms = list(self._alarms.values())
+        return iter(alarms)
+
+    def get(self, alarm_id: str) -> Alarm | None:
+        """Get an Alarm instance."""
+        return self._alarms.get(alarm_id)
+
+    @callback
+    def async_refresh(self, event: SonosEvent, soco: SoCo) -> None:
+        """Create a task to update alarms from an event callback."""
+        if not (update_id := event.variables.get("alarm_list_version")):
+            return
+        if update_id in self._processed_events:
+            return
+        self._processed_events.append(update_id)
```

Comment:
do we still need the processed events? I thought there is now only one speaker.

Reply:
The events from every speaker are dispatched to the new class for handling. This is used to filter out duplicates.

Reply:
We use the `soco` instance from the speaker which fired the event, but only the first one. An alarm change will fire an identical event from every speaker.